File: 35.search-insert-position.java
/*
 * @lc app=leetcode id=35 lang=java
 *
 * [35] Search Insert Position
 */

// @lc code=start
class Solution {
    public int searchInsert(int[] nums, int target) {
        int left = 0; // Khởi tạo chỉ số trái
        int right = nums.length - 1; // Khởi tạo chỉ số phải
        while (left <= right) { // Khi chỉ số trái không vượt quá chỉ số phải
            int mid = left + (right - left) / 2; // Tính chỉ số giữa
            if (nums[mid] == target) { // Nếu phần tử giữa bằng target
                return mid; // Trả về chỉ số giữa
            } else if (nums[mid] < target) { // Nếu phần tử giữa nhỏ hơn target
                left = mid + 1; // Di chuyển chỉ số trái sang phải
            } else { // Nếu phần tử giữa lớn hơn target
                right = mid - 1; // Di chuyển chỉ số phải sang trái
            }
        }
        return left; // Nếu không tìm thấy, trả về chỉ số chèn
    }
}
// @lc code=end

File: 35.search-insert-position.py
#
# @lc app=leetcode id=35 lang=python3
#
# [35] Search Insert Position
#

# @lc code=start
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        if not nums:
            return -1
        low, high = 0, len(nums) - 1
        while low <= high:
            middle = low + (high - low) // 2
            if nums[middle] == target: 
                return middle
            elif nums[middle] < target:
                low = middle + 1
            else:
                high = middle - 1
        return low #kh
# @lc code=end

